### PR TITLE
Add Infrastructure for Geomechanical Simulations and Formation Damage

### PR DESCRIPTION
**Pull request by bska**
This PR adds logic for modifying a connection's transmissibility factor as a result of a fracturing process.  The new member function `Schedule::modifyCompletions()` loops a sequence of externally specified `Connection` objects and incorporates those into the existing set either through altering the CTF directly or through adding a new connection to a well's set of `WellConnections`.  The effect is as if there was a `COMPDAT` section in an `ACTIONX` block, but in which the set of new connections is determined programmatically instead of set in an input file.

While here, also split a few long lines in nearby code and apply `const` where appropriate.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
I don't have any additional work coming here, so I'm marking the PR as "ready for review" now.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 1]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                            perf_exist = true;
+                            if(eperf.CF()> conn.CF()){
+                                conn.setCF(eperf.CF());
+                            }
+                            //auto ctf = conn.CF() + eperf.CF();
+                            //conn.setCF(ctf);
+                            
+                        }    
+                    }
+                    if(!perf_exist){
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: spacing

File: opm/input/eclipse/Schedule/Schedule.hpp [hunk 2]
```diff
@@ -306,6 +307,8 @@ namespace Opm {
         */
         SimulatorUpdate runPyAction(std::size_t reportStep, const Action::PyAction& pyaction, Action::State& action_state, EclipseState& ecl_state, SummaryState& summary_state);
 
+        SimulatorUpdate modifyCompletions(const std::size_t reportStep, 
+                                        const std::map<std::string,std::vector<Opm::Connection>>& extraConns);
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.hpp]: Opm:: in Opm namespace

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 3]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+        SimulatorUpdate sim_update;
+        this->snapshots.resize(reportStep + 1);
+        auto wells = this->snapshots[reportStep].wells();
+        for (auto& wellmap : wells) {
+            // auto& conns = wellmap.second->getConnections();
+            // auto wname = wellmap.first;
+            auto& conns = wellmap.get().getConnections();
+            const auto& wname = wellmap.get().name();
+            if(!(extraConns.find(wname) == extraConns.end())){
+                for(const auto& eperf : extraConns.at(wname) ){
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: spacing

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 4]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                    bool perf_exist=false;
+                    for(auto& conn: conns){
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: spacing

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 5]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                        bool is_equal = (conn.getI() == eperf.getI()) 
+                                    && (conn.getJ() == eperf.getJ())
+                                    && (conn.getK() == eperf.getK());
+                        if(is_equal){
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: spacing

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 6]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                            perf_exist = true;
+                            if(eperf.CF()> conn.CF()){
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: spacing

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 7]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                                conn.setCF(eperf.CF());
+                            }
+                            //auto ctf = conn.CF() + eperf.CF();
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: please no new commented lines

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 8]
```diff
@@ -1786,6 +1786,82 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
...
+                                        eperf.state(),
+                                        eperf.depth(),
+                                        eperf.ctfProperties(),
+                                        1,//eperf.satTableId(),
+                                        eperf.dir(),
+                                        eperf.kind(),
+                                        sort_value,
+                                        /*default table indx*/false);
+                         
+                     //conns.add(eperf);
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: remove

File: opm/input/eclipse/Schedule/Schedule.cpp [hunk 9]
```diff
@@ -1821,9 +1821,10 @@ File {} line {}.)", pattern, location.keyword, location.filename, location.linen
                             
                         }    
                     }
+
                     if(!perf_exist){
                      sim_update.well_structure_changed = true; 
-                     sim_update.affected_wells.insert(wname);
+                     //sim_update.affected_wells.insert(wname);
```
Comment (akva2) [inline opm/input/eclipse/Schedule/Schedule.cpp]: either we need it or we don't need it.

**Review by akva2 — APPROVED** [review]
Just some cosmetics. I realize some of them may be out of date, I went commit by commit.

**Comment by bska** [comment]
jenkins build this please

**Comment by akva2** [comment]
yes, looks good now.

**Comment by bska** [comment]
> looks good now.

Much appreciated!

**Comment by bska** [comment]
PR approved and build check is green.  I'll merge into master.

**Pull request merged**